### PR TITLE
Review

### PR DESCRIPTION
**Pull request by AbramovGit**

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import xgboost as xgb
+from sklearn.model_selection import train_test_split
+import joblib
+import pandas as pd
+
+filename = 'weights/titanic_survivor_grad_boost_classifier.pkl'
+
+
+def convert_csv_to_arrays(path_to_csv: str, train: bool):
+    original_df = pd.read_csv(path_to_csv, sep=',', header=0)\
```
Comment (notsurreal) [inline main.py]: `sep=',', header=0` are default parameters in `pd.read_csv`, you can skip them the next time😉

File: main.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        .drop('Ticket', axis=1)\
+        .drop('Cabin', axis=1)
+    original_df.interpolate(inplace=True)
```
Comment (notsurreal) [inline main.py:17]: Can you please explain what is `interpolate` and what it does here?

File: main.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+    if train:
+        out_df = original_df['Survived'].copy()
+        in_df = original_df.drop('Survived', axis=1)
+    else:
+        in_df = original_df.copy()
+
+    in_df = in_df
+    in_df['Sex'] = in_df['Sex'].map(lambda x: 1 if x == 'male' else 0)
```
Comment (notsurreal) [inline main.py]: This mapping won't work as expected if there are any NaN values in the `Sex` column.

File: main.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    lb = preprocessing.LabelEncoder()
+    min_max_scaler = preprocessing.MaxAbsScaler()
+
+    in_df['Embarked'] = lb.fit_transform(in_df['Embarked'])
+    in_df['Name'] = lb.fit_transform(in_df['Name'])
```
Comment (notsurreal) [inline main.py]: Each feature (column) usually needs its own preprocessing. For continuous numerical columns you can use sklearn's `StandardScaler`, and for categorical and numerical discrete - `OrdinalEncoder`. `LabelEncoder` is used to encode labels. For each column, you should create a preprocessor from scratch, fit it and transform the column.

Here you use the same processor object for different columns, this can cause unexpected preprocessing results.

File: main.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    # in_df['Cabin'] = lb.fit_transform(in_df['Cabin'])
+    in_array = in_df.drop('PassengerId', axis=1).to_numpy()
+    if train:
+        out_array = out_df.to_numpy()
+        return out_array, in_array
+    return in_array, original_df['PassengerId']
+
+
+if __name__ == '__main__':
+    out_array, in_array = convert_csv_to_arrays('data/train.csv', True)
```
Comment (notsurreal) [inline main.py]: I suggest using commonly used notation for input features (`X`) and labels (`y`) and in a different order - features first, then labels, i.e.
```suggestion
    X, y = convert_csv_to_arrays('data/train.csv', True)
```

File: main.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+    in_df['Embarked'] = lb.fit_transform(in_df['Embarked'])
+    in_df['Name'] = lb.fit_transform(in_df['Name'])
+    # in_df['Ticket'] = lb.fit_transform(in_df['Ticket'])
+    # in_df['Cabin'] = lb.fit_transform(in_df['Cabin'])
+    in_array = in_df.drop('PassengerId', axis=1).to_numpy()
+    if train:
+        out_array = out_df.to_numpy()
+        return out_array, in_array
+    return in_array, original_df['PassengerId']
```
Comment (notsurreal) [inline main.py]: Can you please explain what is returned here? Is it for testing?

File: main.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+                  'max_depth': [2, 4, 6, 8, 10, None],
+                  'max_features': [None, 1.0],
+                  'n_estimators': [50, 100, 200],
+                  'min_impurity_decrease': [0.0, 0.1],
+                  'validation_fraction': [0.1, 0.2],
+                  'random_state': [0, None]
+                  }
+    clf = model_selection.GridSearchCV(ensemble.GradientBoostingClassifier(), scoring='roc_auc', param_grid=param_grid)
+    clf.fit(x_train, y_train)
+    print(f'Accuracy is:           {100 * clf.score(x_test, y_test):.2f}%\n')
```
Comment (notsurreal) [inline main.py]: Interesting to see other performance metrics too, e.g. precision, recall, f1-score, and confusion matrix.

File: main.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+                  'learning_rate': [0.1, 0.2, 0.3],
+                  'max_depth': [2, 4, 6, 8, 10, None],
+                  'max_features': [None, 1.0],
+                  'n_estimators': [50, 100, 200],
+                  'min_impurity_decrease': [0.0, 0.1],
+                  'validation_fraction': [0.1, 0.2],
+                  'random_state': [0, None]
+                  }
+    clf = model_selection.GridSearchCV(ensemble.GradientBoostingClassifier(), scoring='roc_auc', param_grid=param_grid)
+    clf.fit(x_train, y_train)
```
Comment (notsurreal) [inline main.py]: Please check out [here](https://scikit-learn.org/stable/auto_examples/model_selection/plot_learning_curve.html) how to plot learning curves for your trained model. Plotting learning curves can help in understanding whether your model is underfit/overfit.

**Review by notsurreal — COMMENTED** [review]
Thank you for your work! You are doing great! I left a few comments to help you improve your project.

File: main.py (hunk 2 above)
Comment (AbramovGit) [reply main.py:17]: Approximate NaN values

File: main.py (hunk 6 above)
Comment (AbramovGit) [reply main.py]: X and passenger ids

File: main.py (hunk 3 above)
Comment (AbramovGit) [reply main.py]: It map 'male' string to 1 and to 0 in the case of not

File: main.py (hunk 1 above)
Comment (notsurreal) [reply main.py]: `\` is not a preferable way to break a line. You can easily do this after opening brackets or after comma, e.g.
```suggestion
    original_df = pd.read_csv(path_to_csv, sep=',', 
                              header=0).drop('Ticket', axis=1)
```

File: main.py (hunk 2 above)
Comment (notsurreal) [reply main.py:17]: Can you please explain why you choose to interpolate NaN values instead of dropping them?

File: main.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+    in_df['Embarked'] = label_encoder.fit_transform(in_df['Embarked'])
+    in_df['Name'] = names_encoder.fit_transform(in_df['Name'].values.reshape(-1, 1))
+    in_df['Cabin'] = cabin_encoder.fit_transform(in_df['Cabin'].values.reshape(-1, 1))
+    print(in_df.head(5))
+    X = in_df.drop('PassengerId', axis=1).to_numpy()
+    if train:
+        y = out_df.to_numpy()
+        return y, X
+    return X, original_df['PassengerId']
```
Comment (notsurreal) [inline main.py:49]: Returning variables in a different order is confusing here. Can you please return variables in the same order (`X, y` or `y, X`) for both train and test?

File: main.py (hunk 3 above)
Comment (notsurreal) [reply main.py]: I understand the logic. My comment is a reminder of possible problems with NaN values.

File: main.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+    in_df = in_df
+    in_df['Sex'] = in_df['Sex'].map(lambda x: 1 if x == 'male' else 0)
+    label_encoder = preprocessing.LabelEncoder()
+    names_encoder = preprocessing.OrdinalEncoder()
+    cabin_encoder = preprocessing.LabelEncoder()
+
+    in_df['Embarked'] = label_encoder.fit_transform(in_df['Embarked'])
+    in_df['Name'] = names_encoder.fit_transform(in_df['Name'].values.reshape(-1, 1))
+    in_df['Cabin'] = cabin_encoder.fit_transform(in_df['Cabin'].values.reshape(-1, 1))
```
Comment (notsurreal) [inline main.py]: I see preprocessing of only three columns. Do you think the rest of the columns are ready to use?

File: main.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+        "std_display_style": "fill_between",
+        "score_name": "Accuracy",
+    }
+    for ax_idx, estimator in enumerate([gb_clf, rf_clf]):
+        LearningCurveDisplay.from_estimator(estimator, **common_params, ax=ax[ax_idx])
+        handles, label = ax[ax_idx].get_legend_handles_labels()
+        ax[ax_idx].legend(handles[:2], ["Training Score", "Test Score"])
+        ax[ax_idx].set_title(f"Learning Curve for {estimator.__class__.__name__}")
+
+    plt.show()
```
Comment (notsurreal) [inline main.py]: `plt.show()` is nice for debugging or to use in notebooks. In a final script, I would recommend saving figures with `plt.savefig()`.

**Review by notsurreal — APPROVED** [review]
This is a good first ml project, keep going!

File: main.py (hunk 7 above)
Comment (notsurreal) [reply main.py]: Have you tried to check those metrics?

File: main.py (hunk 2 above)
Comment (AbramovGit) [reply main.py:17]: Because most of the records would be dropped in this case. Most of the records contains NaN values

File: main.py (hunk 3 above)
Comment (AbramovGit) [reply main.py]: What is your suggestion?
